qiskit/pulse/transforms/resolved_frame.py

```diff
@@ -0,0 +1,210 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Implements a Frame."""
+
+from abc import ABC
+from typing import Optional
+import numpy as np
+
+from qiskit.circuit.parameterexpression import ParameterExpression
+from qiskit.pulse.channels import PulseChannel
+from qiskit.pulse.frame import Frame
+from qiskit.pulse.schedule import Schedule
+from qiskit.pulse.instructions.frequency import SetFrequency, ShiftFrequency
+from qiskit.pulse.instructions.phase import SetPhase, ShiftPhase
+from qiskit.pulse.exceptions import PulseError
+
+
+class Tracker(ABC):
+    """
+    Implements a class to keep track of the phase and frequency of a frame
+    or a pulse channel in a given schedule.
+    """
+
+    def __init__(self, identifier: str, sample_duration: float):
+        """
+        Args:
+            identifier: The identifier of the Tracker.
+            sample_duration: Duration of a sample.
+        """
+        self._identifier = identifier
+        self._frequencies_phases = []  # List of (time, frequency, phase) tuples
+        self._instructions = {}
+        self._sample_duration = sample_duration
+
+    @property
+    def identifier(self) -> str:
+        """Return the identifier of the tracker."""
+        return self._identifier
+
+    def frequency(self, time: int) -> float:
+        """
+        Get the most recent frequency of the tracker before time.
+
+        Args:
+            time: The maximum time for which to get the frequency.
+
+        Returns:
+            frequency: The frequency of the frame right before time.
+        """
+        frequency = self._frequencies_phases[0][1]
+        for time_freq in self._frequencies_phases:
+            if time_freq[0] <= time:
+                frequency = time_freq[1]
+            else:
+                break
+
+        return frequency
+
+    def phase(self, time: int) -> float:
+        """
+        Get the most recent phase of the tracker.
+
+        Args:
+            time: The maximum time for which to get the phase.
+
+        Returns:
+             phase: The phase of the frame up until time.
+        """
+        if len(self._frequencies_phases) == 0:
+            return 0.0
+
+        phase = self._frequencies_phases[0][1]
+        last_time = self._frequencies_phases[0][0]
+
+        for time_freq in self._frequencies_phases:
```

Comment:
I think it's better to use reverse order.

Reply:
See here: b7e8b40

Reply:
Done here: b7e8b40